### PR TITLE
FSE Navigation Sidebar: Move navigation sidebar in DOM hierarchy

### LINKED ISSUE #25701: Site Editor: Navigation panel placement
As pointed out in https://github.com/WordPress/gutenberg/pull/25622#issuecomment-699091615 by @shaunandrews, we should investigate whether it would make sense to render `navigation-panel` as the first element within the the `interface-skeleton`, and potentially moving `navigation-toggle__button` outside of `skeleton__header` too.

This might help with transition animations, tab ordering, and fixing alignment issues of other elements in the header.

### PR DESCRIPTION
**Pull request by jeyip**
## Description

The focus order for the navigation panel wasn't behaving as expected. Shift+Tab from the back button did not focus the "Toggle Navigation" button, but rather the last element of the header toolbar instead.

Although consistent with the Inserter's behavior, Shaun has identified inserter behavior as a bug ([source](https://github.com/WordPress/gutenberg/pull/25622#issuecomment-699091615)). The goal is to update the placement of the navbar toggle and panel so that focus navigation behaves as expected.

## How has this been tested?

**Site Editor**
1. Set up the site editor locally
2. Navigate to the site editor
3. Click on the navigation sidebar toggle
4. Smoke test the navigation sidebar
    - Navigation panel closes correctly when toggle is used again
    - Dashboard button works correctly
    - Navigation panel items behave as expected (click on templates and template parts)
    - Styling matches what is in production
    - Responsiveness matches what is in production
    - After clicking on the navigation toggle, user can press the Shift + Tab keys to refocus the navigation toggle

**Post Editor**
1. Navigate to the post editor
2. Smoke test the page layout:
  - Page styling matches what is in production
  - Page layout matches what is in production
  - General responsiveness matches what is in production
  - **Header**
      - Is the header responsive? 
      - Is the layout correct? 
      - Is it fixed to the screen?
  - **Body**
      - Is the body responsive? 
      - Can the block editor scroll?
      - Is the layout correct? 
  - **Footer**
      - Is the header responsive? 
      - Is the layout correct? 
      - Is it fixed to the screen?
  - **Left Sidebar (e.g. block inserter)**
      - Is the layout correct? 
      - Is it fixed to the screen?
  - **Right Sidebar (e.g. block settings)**
      - Is the layout correct? 
      - Is it fixed to the screen?

## Screenshots 
| Navigation Toggle | Navigation Panel |
|---|---|
| Screen Shot 2020-10-06 at 4 49 05 PM | Screen Shot 2020-10-06 at 4 49 11 PM |

## Types of changes

Bug fix #25701

## Checklist:
- [ ] My code is tested.
- [ ] My code follows the WordPress code style. 
- [ ] My code follows the accessibility standards. 
- [ ] My code has proper inline documentation. 
- [ ] I've included developer documentation if appropriate. 
- [x] I've updated all React Native files affected by any refactorings/renamings in this PR.

### CODE REVIEW

**Comment by jeyip** [comment]
**Note:**

This PR is still a work in progress. When the navigation panel is open, it doesn't shift sibling content correctly.

File: packages/interface/src/components/interface-skeleton/index.js
Comment (jeyip) [inline packages/interface/src/components/interface-skeleton/index.js:67]: **Note:**
Previously, the navigation panel and navigation toggle lived within skeleton__body and skeleton__header, respectively. In order to address unexpected keyboard focus order between the nav toggle and nav panel, we create a new element inside of interface-skeleton called skeleton__navigation-sidebar.

To account for the layout changes of introducing a new element, we add two wrapper classes: skeleton__site and skeleton__content. Both ensure that the skeleton__header and skeleton__body shrink their children correctly when the navigation sidebar is open.

File: packages/interface/src/components/interface-skeleton/index.js
Comment (jeyip) [inline packages/interface/src/components/interface-skeleton/index.js:66]: Previously, the navigation panel and navigation toggle lived within skeleton__body and skeleton__header, respectively. In order to address unexpected keyboard focus order between the nav toggle and nav panel, we create a new element inside of interface-skeleton called skeleton__navigation-sidebar.

To account for the layout changes of introducing a new element, we add two wrapper classes: skeleton__site and skeleton__content. Both ensure that the skeleton__header and skeleton__body shrink their children correctly when the navigation sidebar is open.

File: packages/edit-site/src/components/navigation-sidebar/navigation-toggle/style.scss
Comment (jeyip) [inline packages/edit-site/src/components/navigation-sidebar/navigation-toggle/style.scss:9]: **Note:**
When the navigation sidebar is closed, the skeleton body (i.e. the block editor) no longer spans the entire width of the viewport. The sidebar (and toggle) occupy space at the left edge of the screen. To address this, we position the navigation toggle absolutely when the sidebar is closed.

File: packages/interface/src/components/interface-skeleton/style.scss
Comment (jeyip) [inline packages/interface/src/components/interface-skeleton/style.scss:148]: **Note:**

The sidebar border was overlapping the footer, which is why we set an explicit z-index of 90. I'm not certain where this originates from, but I'll be looking more into it.

Comment (david-szabo97) [reply packages/interface/src/components/interface-skeleton/style.scss:148]: I just found this file: https://github.com/WordPress/gutenberg/blob/master/packages/base-styles/_z-index.scss  See line 83-84-85

Might be helpful for you and we could add this to the mapping with a comment explaining the value.

Comment (Copons) [reply packages/interface/src/components/interface-skeleton/style.scss:148]: Good point @david-szabo97, and it's a good reminder to use the [`z-index`](https://github.com/WordPress/gutenberg/blob/a8b872408e751b78ad9790ddc1f0446c0577e32e/packages/base-styles/_z-index.scss#L143-L149) mixin (and add the key to the list) when defining new z-indices.

File: packages/edit-site/src/components/navigation-sidebar/index.js
Comment (Copons) [inline packages/edit-site/src/components/navigation-sidebar/index.js:15]: Should we keep the `MainDashboardButton.Slot` removed from `packages/edit-site/src/components/header/index.js`?

**Comment by Copons** [comment]
The behaviour here feels good to me, and the keyboard navigation is exactly as I would expect.

As mentioned elsewhere, I'd even go as far as removing the [auto-focus](https://github.com/WordPress/gutenberg/blob/c1390e0bdd09bd73125317a7f2320ca78726e725/packages/edit-site/src/components/left-sidebar/navigation-panel/index.js#L27-L33) on the first button inside the navigation (Back to Dashboard) on sidebar toggle, since now it's only as far as a simple tab press.
I'd defer this decision to a11y, though.

---

Most obvious change required: replace all the new z-index values with the `z-index` mixin, as mentioned in an inline comment.

---

About the `InterfaceSkeleton` change.

I don't see any obvious regressions, but I'm wondering if there's a way to tidy it up a little bit.
(Note that I haven't tested my propositions, so I'm not certain they would work 🙂)

- Is the `interface-interface-skeleton__layout` wrapper essential?
Couldn't we just use the root `interface-interface-skeleton` as container?
Off the top of my head, the reason for `__layout` is so that the `interface-interface-skeleton__footer` is not affected by the navigation sidebar sliding in.
What if, instead, the sidebar "moved" the footer as well, exactly like it moves header and content?
(Note that this is possibly kinda awkward, considering that status bars are usually full-width unaffected by any other UI element of the window.)

- How do you feel about `interface-interface-skeleton__editor` instead of `interface-interface-skeleton__site`?

- Is "navigation sidebar" a good and descriptive name for the new skeleton element?
I wonder if "navigation" is too semantically specific, whereas I guess we should probably just convey an "interface" meaning (e.g. just "sidebar").
For example, "Drawer" sounds like a good and descriptive name to me: it's established as a UI element (e.g. Google's Material), and it conveys a "sliding" movement, different from the other sidebars, who just simply appear.

File: packages/edit-site/src/components/navigation-sidebar/index.js
Comment (vindl) [reply packages/edit-site/src/components/navigation-sidebar/index.js:15]: I think it should now wrap `Back to Dashboard` button, instead of the main one that toggles the sidebar. That's what it does in post editor too.

**Comment by jeyip** [comment]
> Is the interface-interface-skeleton__layout wrapper essential?

Great catch! You were spot on about why I initially implemented `skeleton-layout`, but it turns out that we can remove the layout wrapper while still preserving current behavior. `skeleton-footer` has position `fixed`, which always attaches it to the bottom of the screen 🙂

> How do you feel about interface-interface-skeleton__editor instead of interface-interface-skeleton__site?

`editor` is more semantically specific but it conveys a lot more meaning. I'm on board.

> Is "navigation sidebar" a good and descriptive name for the new skeleton element?
I wonder if "navigation" is too semantically specific, whereas I guess we should probably just convey an "interface" meaning (e.g. just "sidebar"). For example, "Drawer" sounds like a good and descriptive name to me: it's established as a UI element (e.g. Google's Material), and it conveys a "sliding" movement, different from the other sidebars, who just simply appear.

Agreed. I'll give this more thought and circle back tomorrow. For now, I'll use drawer instead of navigation-sidebar.

File: packages/edit-site/src/components/editor/index.js
Comment (Copons) [inline packages/edit-site/src/components/editor/index.js:213]: (More of a pondering than a request)

Given that now the navigation sidebar is completely separated from the left sidebar, I think they should be handled by different states (e.g. `setLeftSidebar` and `setNavigationSidebar`).

With different states we could even have both sidebar visible at the same time.
I don't know if it's something we want, though, as we might end up with 3 sidebars open at the same time, which feels very cluttered to me.

Either way, we can make navigation and left sidebars "interdependent": when you open one, automatically close the other.

File: packages/edit-site/src/components/navigation-sidebar/navigation-panel/style.scss
Comment (Copons) [inline packages/edit-site/src/components/navigation-sidebar/navigation-panel/style.scss:3]: (Not caused by this PR, but it's more a mental note to address it at some point)

The panel is 100% height, but in fact it should be slightly shorter: it's pushed down by the header (60px + 1px of border), and it's supposed to end above the footer (25px, only on "> mobile").

I guess we could remove those 86px from the height with `calc()` or as a bottom padding.

Somewhat related: the Navigation component hides both overflows, while it only really needs to hide `overflow-x`.
Hiding `overflow-y` means that if the menu is taller than the screen, it gets cut off.

File: packages/edit-site/src/components/editor/index.js
Comment (Copons) [inline packages/edit-site/src/components/editor/index.js:204]: I think we might want to add a custom label for the drawer in `interfaceLabels`.

Like, the interface element is called "Drawer" but it's filled by "Navigation Sidebar" (or at least I think that's the point of `interfaceLabels` 😄).

**Review by Copons — COMMENTED** [review]
This is looking good to me!

I've left a few comments: all minor stuff and discussion points, possibly to keep in mind for follow ups!

I'd love to have a few more 👀 on the interface changes, as this is the first time I've dabbled with it.

File: packages/edit-site/src/components/editor/index.js
Comment (jeyip) [reply packages/edit-site/src/components/editor/index.js:204]: Good catch!

File: packages/edit-site/src/components/editor/index.js
Comment (noahtallen) [reply packages/edit-site/src/components/editor/index.js:214]: I wanted to put the nav state into redux so I could use it in a random component, so i tried this out here: https://github.com/WordPress/gutenberg/pull/26003

**Comment by vindl** [comment]
This is working as expected for me. The PR now needs a rebase before we can merge it.

**Comment by shaunandrews** [comment]
Very cool. Keyboard focus moves as I expected.

--

Maybe not entirely related to this PR, but the transition feels a little wonky. It feels strange as the other canvas and top toolbar aren't being transitioned.

**Comment by jeyip** [comment]
> This is working as expected for me. The PR now needs a rebase before we can merge it.

@vindl Apologies for the delay! Rebasing it now.

**Comment by jeyip** [comment]
> Very cool. Keyboard focus moves as I expected.

🎉 

> Maybe not entirely related to this PR, but the transition feels a little wonky. It feels strange as the other canvas and top toolbar aren't being transitioned.

Good observation. I think it may be caused by this PR, but I'm not sure either. I'll take a look.

**Comment by jeyip** [comment]
> Maybe not entirely related to this PR

@shaunandrews Yeah I don't think it's this PR in particular. I'm not seeing transitions on the master branch. I agree that the animation feels wonky when everything else jumps. How would we feel about creating a follow-up issue to be prioritized accordingly?

**Comment by Copons** [comment]
This looks good to me!
@jeyip I think we should wait for #26003 to merge so that we can use the Redux state for the sidebars.

I'll also create two new issues to tackle the style issues emerged from this PR:

- Review the Nav sidebar animation in relation to the rest of the content.
Currently: on mount, the sidebar starts "outside" the screen, but the rest of the content immediately repositions to make room for the sidebar. I guess everything should move at the same time (sidebar and nav toggle sliding in from the left; content gradually decreasing width).

- See https://github.com/WordPress/gutenberg/pull/25884#discussion_r502370320
  - The Nav _component_ should only hide the `overflow-x`. Hiding the `overflow-y` means that if the content is > the height, its cut off.
  - The Nav _sidebar_ should calculate its height, subtracting the header and footer (except on mobile, where the footer is hidden).

**Comment by jeyip** [comment]
> This looks good to me!
@jeyip I think we should wait for #26003 to merge so that we can use the Redux state for the sidebars.

Sounds good! Let me know if you need any help / want me to draft the two issues you mentioned. I'm happy to help 🙂

**Comment by Copons** [comment]
> Sounds good! Let me know if you'd like any assistance / want me to draft either or both of the two issues you mentioned. I'm happy to help 🙂

Apologize, I got stuck in various rabbit holes, but I've opened the issues now! #26119 #26120

File: packages/edit-site/src/components/header/index.js
Comment (noahtallen) [inline packages/edit-site/src/components/header/index.js:33]: Looks like we can remove this since we aren't using it any more

File: packages/edit-site/src/components/header/index.js
Comment (noahtallen) [inline packages/edit-site/src/components/header/index.js:47]: same as above (can be removed)

File: packages/edit-site/src/components/navigation-sidebar/index.js
Comment (noahtallen) [inline packages/edit-site/src/components/navigation-sidebar/index.js:15]: we can simplify this slightly by returning the value directly:


```js
const isNavigationOpen = useSelect( (select) => {
	return select( 'core/edit-site' ).isNavigationOpened();
} );
```

File: packages/edit-site/src/components/navigation-sidebar/navigation-toggle/index.js
Comment (noahtallen) [inline packages/edit-site/src/components/navigation-sidebar/navigation-toggle/index.js:9]: since we no longer pass the `onClick` handler through, should we also get `isOpen` from the state directly here instead of via props? (not sure myself, up for discussion)

**Review by noahtallen — COMMENTED** [review]
This works well for me in my testing, nice work! I added a few nitpicky comments re. the redux state integration. 

Some thoughts on keyboard navigation:
- should `esc` close the sidebar (and inserter sidebar for that matter)
- In the quick inserter, tab moves between the different sections, and then arrow keys move between blocks. Should we let tab toggle between top bar and nav panel, and then arrow keys to move between actions in the nav panel? Unsure what best practice is here.

File: packages/edit-site/src/components/header/index.js
Comment (jeyip) [reply packages/edit-site/src/components/header/index.js:33]: Good catch

File: packages/edit-site/src/components/navigation-sidebar/navigation-toggle/index.js
Comment (jeyip) [reply packages/edit-site/src/components/navigation-sidebar/navigation-toggle/index.js:9]: `NavigationToggle`'s direct parent, `NavigationSidebar` also uses the isOpen prop (which resolves to isNavigationOpened). It felt odd having identical selectors in both the parent and child, but I don't have a strong opinion here.

**Review by Copons — APPROVED** [review]
Looking good! Thanks @jeyip! ✨

**Review by david-szabo97 — APPROVED** [review]

**Comment by david-szabo97** [comment]
Sorry! Those two commits are meant to be committed to a different branch 😬

**Comment by Copons** [comment]
> Sorry! Those two commits are meant to be committed to a different branch 😬

No worries! Rebased the PR and dropped the rogue commits.

**Comment by jeyip** [comment]
@Copons Thanks for merging this! 🙏

File: packages/base-styles/_variables.scss
Comment (jasmussen) [inline packages/base-styles/_variables.scss:59]: @jeyip I know this value is iPhone friendly, but it doesn't appear to work as intended as the mobile experience is otherwise broken. The variable is also used only in this PR and no-where else. I'd recommend not adding this variable, and instead either use `$break-mobile` from _breakpoints.scss so as to avoid fragmenting the responsive dev.

We also have some good mixins that help mobile dev, such as `@include break-small() { ... }` to abstact and keep things in sync.

**Comment by jasmussen** [comment]
This PR appears to consider mobile (specificaly the iPhone based on the 335px variable introduced), but the mobile experience remains broken, so I'm curious how this was tested for mobile. For me, the site editor is just a white screen on mobile, something which #26021 fixes, but even after that fix, the 335px variable appears to not really address any mobile issues. 

I really appreciate all the work you're doing on this, and I know it's super early to optimize for a breakpoint that is likely going to be rarely used for building templates. However this will nevertheless need to be considered at some point, and the sooner we start testing it, the less technical debt we will have to revisit and fix. 

Most of the time it's it's enough to wrap desktop-specific affordances in a `@include break-small() { ... }` clause.

**Comment by jeyip** [comment]
> This PR appears to consider mobile (specificaly the iPhone based on the 335px variable introduced), but the mobile experience remains broken, so I'm curious how this was tested for mobile.

Thanks for your thoughts @jasmussen. They're much appreciated.

To clarify, although the PR appears to consider mobile (like the 335px variable), I didn't explicitly take mobile screens into account, which admittedly might be an oversight. I only ensured that the desktop behaved responsively up to the point where the tablet / mobile break-points take over. I've seen the mobile white screen for a while in site editor development, which is part of the reason why I didn't test behavior at smaller screen widths. I should've, however, been more proactive about calling out the unexpected behavior that you're addressing in #26021.

> I really appreciate all the work you're doing on this, and I know it's super early to optimize for a breakpoint that is likely going to be rarely used for building templates. However this will nevertheless need to be considered at some point, and the sooner we start testing it, the less technical debt we will have to revisit and fix.

I think these are very valid points, especially if we plan on supporting mobile site editing in the future. It's tough developing for a mobile experience without an idea of what it will look like, but I agree that we can circumvent a lot of issues with desktop-specific affordances wrapped in breakpoint mixins.

File: packages/base-styles/_variables.scss
Comment (jeyip) [reply packages/base-styles/_variables.scss:59]: The 335px was meant to ensure a certain width for the toolbar on either side of the header used [here](https://github.com/WordPress/gutenberg/pull/25884/files/4a2937aca4377f3eda568b93d865980e7e72f893#diff-0cda043c56d581dee157884a2abbedb914c3e24d9712cf0317fea8de5a4bf1a9R16-R20) and [here](https://github.com/WordPress/gutenberg/pull/25884/files/4a2937aca4377f3eda568b93d865980e7e72f893#diff-0cda043c56d581dee157884a2abbedb914c3e24d9712cf0317fea8de5a4bf1a9R34-R40). The min width prevented the toolbars from collapsing as the browser viewport was shrunk horizontally. Giving this more thought, it should absolutely be a desktop specific affordance. 

Screen Shot 2020-10-26 at 3 21 32 PM

I'll open another PR incorporating the break point mixins.

Comment (jasmussen) [reply packages/base-styles/_variables.scss:59]: Ah! Thank you for the explanation. I initially pondered where those 335px came from, because it seems a relatively arbitrary number given the base8 grid most of the UI uses. I then thought — maybe it's that somewhat odd iphone resolution which I _remembered_ as 335px (iPhone 6: 750 x 1334, which makes it 375 points horizontally) but as you can see, I was wrong.

With your explanation, though, I think a different approach is needed for min-width. Longer term I think we need to stack the toolbar if there isn't space, but nearer term I think smaller steps can be taken.

> I'm happy to open another PR incorporating the break point mixins.

#26021 does that, and but also hides some buttons, and minimizes labels, as the viewport approaches mobile breakpoints. It's only a first step, and more are needed, but it's a good step. It's currently blocked on the white screen as shown in the intro of the PR, which happens because the left sidebar is always output, regardless of open or not. I could use help with that if you have bandwidth!

> I'm not quite sure I'm tracking here. I think the name of the variable may be misleading, because it sounds like we see the 335px as an iphone friendly breakpoint?

To be specific, we should only be creating new variables in the _variables.scss file if they are meant to be used widely and in multiple places. If it's just for the site editor as is currently the case, I'd just type the content of the variable directly where needed, or just register the variable directly in those files.

I'd look at finding a different approach than the min-width, though (not urgently) it seems a half-step towards solving a problem that needs a more foundational mobile improvement. The work could probably coincide with enforcing of the stacking top toolbar, so we fix this:

Screenshot 2020-10-27 at 09 55 55

I know it's beta! I know there's a lot of work to be done. I'm just:

- excited about the site editor
- of the belief that if we test mobile in every PR, even if it's not superb, we minimize future headaches

**Comment by jasmussen** [comment]
>  It's tough developing for a mobile experience without an idea of what it will look like, but I agree that we can circumvent a lot of issues with desktop-specific affordances wrapped in breakpoint mixins.

Absolutely.

And in absence of full knowledge of that, I'd just test site editor PRs against a viewport that's <600px wide and see how broken it is. As we move forward, we can then raise the bar for what brokenness we're willing to accept.

File: packages/base-styles/_variables.scss
Comment (jeyip) [reply packages/base-styles/_variables.scss:59]: I added a [commit](https://github.com/WordPress/gutenberg/pull/26021/commits/b4746b5bc04e676134df1a714d51896fff399668) and [a comment](https://github.com/WordPress/gutenberg/pull/26021#issuecomment-717095993) to #26021 that potentially address the issues you bring up here.

**Pull request merged**
